### PR TITLE
BKNDLSS-31061 Mapbox

### PR DESCRIPTION
**Pull request by addKiwi**

### CODE REVIEW

File: components/bl-mapbox-component/src/helpers.js [hunk 1]
```diff
@@ -121,64 +121,77 @@ export const useMarkers = (markers, markersArray, setMarkersArray, mapRef, onMar
...
-          'type': 'geojson',
-          'data': {
-            'type'    : 'Feature',
-            'geometry': {
-              'type'       : 'Polygon',
-              'coordinates': [[...polygon.coordinates]],
-            },
+const addPolygons = (mapRef, polygonsArray, onPolygonClick) => {
+  mapRef.current.on('load', () => {
+    polygonsArray.forEach(polygon => {
```
Comment (v-excelsior) [inline components/bl-mapbox-component/src/helpers.js:141]: const { description, id, coordinates ... } = polygon

File: components/bl-mapbox-component/src/helpers.js [hunk 2]
```diff
@@ -121,64 +121,77 @@ export const useMarkers = (markers, markersArray, setMarkersArray, mapRef, onMar
...
-          }
-        });
+const updatePolygonsArray = (polygons, mapRef, polygonsArray, setPolygonsArray) => {
+  if (polygons?.length && mapRef.current) {
+    mapRef.current.on('load', () => {
+      polygonsArray.forEach(polygon => {
+        if (mapRef.current.getSource(polygon)) {
+          mapRef.current.removeLayer(`${ polygon.id }-layer`);
+          mapRef.current.removeSource(polygon.id);
+        }
```
Comment (v-excelsior) [inline components/bl-mapbox-component/src/helpers.js:131]: Can we check if the polygon changed and remove then only if it is?

File: components/bl-mapbox-component/src/helpers.js [hunk 3]
```diff
@@ -121,64 +121,77 @@ export const useMarkers = (markers, markersArray, setMarkersArray, mapRef, onMar
...
-          },
-        });
+      mapRef.current.addLayer({
+        'id'    : `${ polygon.id }-layer`,
+        'type'  : 'fill',
+        'source': polygon.id,
+        'layout': {},
+        'paint' : {
+          'fill-color'  : polygon.color,
+          'fill-opacity': polygon.opacity,
```
Comment (v-excelsior) [inline components/bl-mapbox-component/src/helpers.js:160]: Can we avoid using notation like 'id' for objects? As for - 'fill-color'. It doesn't work if we use camelCase('fillColor')?

File: components/bl-mapbox-component/src/helpers.js [hunk 4]
```diff
@@ -121,64 +121,77 @@ export const useMarkers = (markers, markersArray, setMarkersArray, mapRef, onMar
...
 
-            const coordinates = { lat: e.lngLat.lat, lng: e.lngLat.lng };
+          const coordinates = { lat: e.lngLat.lat, lng: e.lngLat.lng };
 
-            onPolygonClick({ coordinates: coordinates, description: polygon.description });
-          });
-        } else {
-          mapRef.current.on('click', `${ polygon.id }-layer`, e => {
-            const coordinates = { lat: e.lngLat.lat, lng: e.lngLat.lng };
+          onPolygonClick({ coordinates: coordinates, description: polygon.description });
```
Comment (v-excelsior) [inline components/bl-mapbox-component/src/helpers.js:173]:
```suggestion
          const {description, ... } = polygon
          
          ...
          
          
          onPolygonClick({ coordinates, description });
```

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-mapbox-component/src/helpers.js (hunk 2 above)
Comment (addKiwi) [reply components/bl-mapbox-component/src/helpers.js:131]: We are already checking it in usePolygons function.

File: components/bl-mapbox-component/src/helpers.js (hunk 3 above)
Comment (addKiwi) [reply components/bl-mapbox-component/src/helpers.js:160]: Every polygon needs a unique id to be added on map, I think it is not a good idea to ask users to create a unique name(id) for every polygon. 'fill-color' and 'fill-opacity' doesn't work if we use camelCase.

File: components/bl-mapbox-component/src/helpers.js [hunk 5]
```diff
@@ -1,9 +1,29 @@
...
+const mapOnLoad = (mapRef, listener) => {
+  mapRef.current.on('load', listener);
+};
+
+const mapOnClick = (mapRef, layerId, listener) => {
+  if (layerId) {
+    mapRef.current.on('click', layerId, listener);
+  } else {
+    mapRef.current.on('click', listener);
+  }
```
Comment (v-excelsior) [inline components/bl-mapbox-component/src/helpers.js:16]: const handlerPayload = layerId ? [layerId, listener] : [listener]

mapRef.current.on('click', ...handlerPayload);

File: components/bl-mapbox-component/src/helpers.js [hunk 6]
```diff
@@ -121,70 +145,92 @@ export const useMarkers = (markers, markersArray, setMarkersArray, mapRef, onMar
...
+      mapAddLayer(mapRef, {
+        'id'    : `${ id }-layer`,
+        'type'  : 'fill',
+        'source': id,
+        'layout': {},
+        'paint' : {
+          'fill-color'  : color,
+          'fill-opacity': opacity,
+        },
       });
```
Comment (v-excelsior) [inline components/bl-mapbox-component/src/helpers.js:209]: Did you try Class approach for basic Map functions?

File: components/bl-mapbox-component/src/map.js [hunk 7]
```diff
@@ -1,27 +1,26 @@
...
-  const [polygonsArray, setPolygonsArray] = useState([]);
-  const [markersArray, setMarkersArray] = useState([]);
-
   const { accessToken } = settings;
 
   const { onMarkerClick, onPolygonClick } = eventHandlers;
 
   const { markers, polygons, center, classList } = component;
 
+  const map = new Map(mapRef);
```
Comment (Valodya) [inline components/bl-mapbox-component/src/map.js:18]: it creates a new instance on each render and then in the next useEffect it remembers the first one, which means the `initMapboxLibrary ` runs only for the first instance

File: components/bl-mapbox-component/src/helpers.js [hunk 8]
```diff
@@ -51,21 +96,21 @@ export const initMapboxLibrary = (mapRef, mapContainerRef, component, eventHandl
   createActions(mapRef, component);
 
   if (directions) {
-    mapRef.current.addControl(
+    map.addControl(
```
Comment (Valodya) [inline components/bl-mapbox-component/src/helpers.js:99]: please test changing each property of the component in runtime

**Pull request merged**